docs/design/2021-03-09-dynamic-privileges.md

```diff
@@ -0,0 +1,282 @@
+```
+mysql> grant acdc on *.* to u1;
+ERROR 8121 (HY000): privilege check fail
+
+### Attempting to set DYNAMIC privilege: acdc
+```
+
+It might be possible that changes are still required if `ExtendedPriv` is not supported in all contexts (REVOKE, etc).
+
+Note that creating a role with the same name as a DYNAMIC privilege is supported. A `GRANT` statement can be attributed to ROLES when it omits the ON *.* syntax. Thus:
+
+```
+GRANT BINLOG_ADMIN TO u1; // grants the role binlog_admin
+GRANT BINLOG_ADMIN ON *.* TO u1; // grants the dynamic privilege binlog_admin
+```
+
+This same nuance applies to MySQL.
+
+### Documentation Plan
+
+The statement reference pages for each of the affected metadata commands will need to be updated to describe dynamic privileges.
+
+There will also need to be documentation specific to `DYNAMIC` privileges to describe how it works, and the purpose of fine-grained access control.
+
+
+## Test Design
+
+Testing dynamic privileges is a little bit complex because of the various ways privileges can be inherited by users:
+
+* Direct `GRANT` to the user
+* Granting to a role that the user inherits.
+
+### Functional Tests
+
+Unit tests will be added to cover the semantics around role/dynamic privilege precedence, including logical restoring in a different order.
+
+Integration testing needs to test with global kill enabled/disabled.
+
+### Scenario Tests
+
+The use-cases required by the DBaaS team should be validated when combined with `security-enhanced-mode`. They are:
+
+| Account Name | root | cloudAdmin |
+| --------------- | --------------- | --------------- |
+| Backup & Restore to cloud | Y | Y |
+| File privilege | N | N |
+| Read or set variables | Y | Y | 
+| set restricted variables(some of them even can not be read) | N | Y |
+| Read or set restricted system tables | N | Y | 
+| DROP USER cloudAdmin | N | Hard to N(unless some hardcoded) |
+| REVOKE cloudAdmin | N | Hard to N(unless some hardcoded) |
+| Show processlist, Access to threads belong to other users | Y | Y |
+| Change password if the password of SUPER user is forgotten | N | Y |
+| Kill connections belong to cloudAdmin | N | Y |
+| SHUTDOWN / RESTART | N | Y (graceful shutdown on k8s for tidb-server) |
+
+### Compatibility Tests
+
+The introduction of `DYNAMIC` privileges is not expected to introduce any compatibility issues, because backwards compatibility is ensured. However, plugins should migrate to registering their own dynamic privileges and not rely on the use of `SUPER`. This is considered an enhancement, and not included in-scope for the initial introduction of dynamic privileges (which introduces the framework for plugins to use).
+
+## Impact & Risks
+
+In its initial release, dynamic privilege usage will be controlled by an experimental feature flag. The implementation will be via restricting `GRANT` and `REVOKE` statements from creating dynamic privileges (it is too invasive to conditionally modify the ast visitor functionaliy).
```

Comment:
could you describe more about the feature flag? is it a global variable that can or can't be seen by users via `show variables`?

Reply:
Yes. I added `tidb_enable_dynamic_privileges` in https://github.com/pingcap/tidb/pull/22778/commits/b7809f6fa339539861bbd5ae96ec6730951e456a